### PR TITLE
Logging Output Manager Filtering Activity

### PR DESCRIPTION
**Pull request by ew3361zh**
This PR adds OM logging within OM's own filtering functions.

This tracks vital OM filtering activity including:
- The status of the exclude_info_maps flag for the simulation
- The number of filter files found in the filter files directory
- Whether the exclude_keyword was found in the first position of each particular filter file
- The number of keys resulting from each application of the filter file's patterns to the variables pool

This will be helpful to users and developers diagnose unexpected results found in filtered variables output files.

Also updates unit testing accordingly.

Note:
- Fixes issue of OM only logging one of the new logs by switching the order of OM function calls in main.py

### CODE REVIEW

File: RUFAS/output_manager.py [hunk 1]
```diff
@@ -351,13 +366,20 @@ def _load_txt_file_to_list(self, path: str) -> List[str]:
...
         """
+        info_map = {"class": self.__class__.__name__,
+                    "function": self._load_txt_file_to_list.__name__,
+                    }
+        self.add_log("open_text_file", f"Attempting to open {path}.", info_map)
         try:
             with open(path) as text_file:
-                return text_file.read().splitlines()
+                list_of_elements = text_file.read().splitlines()
+                load_message = f"Successfully opened {path}, it contained {len(list_of_elements)} filter patterns."
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
                load_message = f"Successfully opened {path} and read {len(list_of_elements)} lines."
```
This function is "_load_txt_file_to_list" which means here we don't what are these lines yet. We had the same thing in a previous PR that led to renaming this function to its current name.
Also, at this point, we have successfully read the content of the file, we should log that too. Hence changing 'contained' to 'read'.

File: RUFAS/output_manager.py [hunk 2]
```diff
@@ -366,6 +388,9 @@ def _filter_variables_pool(self, filter_patterns: List[str]) -> Dict[str, pool_e
         filter_patterns : List[str]
             A list of patterns the user has selected to filter the variables pool.
 
+        input_file : str
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
        input_file_name : str, optional
```

File: RUFAS/output_manager.py [hunk 3]
```diff
@@ -351,13 +366,20 @@ def _load_txt_file_to_list(self, path: str) -> List[str]:
...
-                return text_file.read().splitlines()
+                list_of_elements = text_file.read().splitlines()
+                load_message = f"Successfully opened {path}, it contained {len(list_of_elements)} filter patterns."
+                self.add_log("filter_pattern_file_load_log", load_message, info_map)
+                return list_of_elements
         except Exception as e:
             raise e
 
-    def _filter_variables_pool(self, filter_patterns: List[str]) -> Dict[str, pool_element_type]:
+    def _filter_variables_pool(self, filter_patterns: List[str], input_file: str) -> Dict[str, pool_element_type]:
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
    def _filter_variables_pool(self, filter_patterns: List[str], input_file_name: Optional[str]) -> Dict[str, pool_element_type]:
```
1. It is the file name, not the file object itself.
2. The function can work without it, hence optional.

File: RUFAS/output_manager.py [hunk 4]
```diff
@@ -380,15 +405,27 @@ def _filter_variables_pool(self, filter_patterns: List[str]) -> Dict[str, pool_e
...
+                    "function": self._filter_variables_pool.__name__,
+                    }
         exclude_keyword_location = 0
         exclude_keyword = "exclude"
         filter_by_exclusion = filter_patterns and filter_patterns[exclude_keyword_location] == exclude_keyword
         if filter_by_exclusion:
-            return {key: self.variables_pool[key] for key in self.variables_pool.keys() if not
-                    any(re.match(pattern, key) for pattern in filter_patterns)}
+            filter_vars_msg = f"{input_file} has exclude-keyword '{exclude_keyword}' at"\
+                f" position {exclude_keyword_location}."
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
                f" position {exclude_keyword_location}. Performing filtering by exclusion."
```

File: RUFAS/output_manager.py [hunk 5]
```diff
@@ -380,15 +405,27 @@ def _filter_variables_pool(self, filter_patterns: List[str]) -> Dict[str, pool_e
...
+            filter_pattern_matches = {key: self.variables_pool[key] for key in self.variables_pool.keys() if not
+                                      any(re.match(pattern, key) for pattern in filter_patterns)}
         else:
-            return {key: self.variables_pool[key] for key in self.variables_pool.keys() if
-                    any(re.match(pattern, key) for pattern in filter_patterns)}
+            filter_vars_msg = f"{input_file} does NOT contain exclude-keyword '{exclude_keyword}'"\
+                f" at position {exclude_keyword_location}."
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]:
```suggestion
                f" at position {exclude_keyword_location}. Performing filtering by inclusion."
```

File: RUFAS/output_manager.py [hunk 6]
```diff
@@ -408,15 +445,18 @@ def save_variables(self, save_path: str, dir_path: str,
             Flag for whether or not the user wants to include info_maps data in their results files.
 
         """
+        info_map = {"class": self.__class__.__name__,
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py:449]: I feel bad for this info_map, we are about to exclude it :(

**Review by PooyaHekmati — APPROVED** [review]
Please address comments before merging.

File: RUFAS/output_manager.py (hunk 6 above)
Comment (ew3361zh) [reply RUFAS/output_manager.py:449]: Haha, it did make me giggle thinking about that when writing it in there.

**Review by JoeWaddell — APPROVED** [review]
Looks good to me! Nice work, and I always appreciate the detailed comment to open the review.

**Pull request merged**